\documentclass[aps,rmp, onecolumn]{revtex4}
%\documentclass[a4paper,10pt]{scrartcl}
%\documentclass[aps,rmp,twocolumn]{revtex4}

\usepackage[utf8]{inputenc}
\usepackage{amsmath,graphicx}
\usepackage{color}
%\usepackage{cite}

\newcommand{\bq}{\begin{equation}}
\newcommand{\eq}{\end{equation}}
\newcommand{\bn}{\begin{eqnarray}}
\newcommand{\en}{\end{eqnarray}}
\newcommand{\Vadim}[1]{{\color{blue}Vadim: #1}}
\newcommand{\Richard}[1]{{\color{red}Richard: #1}}
\newcommand{\gene}[1]{{\it #1}}
\newcommand{\mat}[1]{{\bf #1}}
\newcommand{\vecb}[1]{{\bf #1}}
\newcommand{\abet}{\mathcal{A}}
\newcommand{\eqp}{p}
\newcommand{\LH}{\mathcal{L}}
\newcommand{\lh}{\ell}
\pdfinfo{%
  /Title    ()
  /Author   ()
  /Creator  ()
  /Producer ()
  /Subject  ()
  /Keywords ()
}


\begin{document}

\title{Site-specific substitution model inference with iterative tree likelihood maximization.}
\author{Vadim Puller$^{1,2,3}$ Pavel Sagulenko$^{1}$, Richard Neher$^{1,2,3}$}
\affiliation{$^{1}$Max Planck Institute for Developmental Biology, 72076 T\"ubingen, Germany\\
$^{2}$Biozentrum, University of Basel, Klingelbergstrasse 50/70, 4056 Basel, Switzerland\\
$^{3}$SIB Swiss Institute of Bioinformatics, Basel, Switzerland}

\date{\today}

\maketitle

\section*{Introduction}
Over time, genome sequences change through mutations and are reshuffeled by recombination.
Modifications to the genomes are filtered by selection for survival such that beneficial variants spread preferentially and those that impair function are rare.
As a result, some parts of genomes change rapidly, while other are strongly conserved.
In addition to variation in rate, different sites in a genome only explore subsets of the available states.
Some positions in a protein, for example, might only allow for hydrophobic amino acids, while other require acidic side chains.

Phylogenetics aims at reconstructing the relationships and history of homologous sequences from the the substitutions that occurred in the past.
Modern phylogenetic methods describe this stochastic evolutionary process with probabilistic models of sequence evolution and aim to find phylogenies that either maximize the likelihood of observing the alignment or sample phylogenies from a posterior probability distribution \citep{felsenstein2004inferring}.

Inferring phylogenies is a computationally challenging problem since (i) the number of phylogenies grows super-exponentially with the number of taxa, (ii) calculation of the likelihood is compulationally costly (though linear in $n$), (iii) substitution models and the tree have many unknown continuous parameters that need estimation.
Here, we will focus on the latter problem and describe ways to estimate substitution models in computationally efficient and economically parameterized form.

The simplest substitution models assume that all sites and sequence states are equivalent and evolve at the same rate, i.e., they assume an unconstrained non-functional sequence that mutates at random between the different sequence states.
Such simple models are clearly inadequate.
But increasing the number of parameters, although improving the likelihood, does not necessarily increases the statistical/predictive power of the model \cite{scheffler2014validity}.
In the context of the site-specific models this issue has become known as {\em extensive parametrization} or even {\em infinitely many parameters problem} \cite{felsenstein2001taking,Rodrigue557,kumar2011statistics,spielman2016extensively}.
%
A popular way to overcome this problem is by using mixture models, where every site is assigned to a category with its substitution rate drawn randomly from a statistical distribution assigned to this category \cite{blanquart2008site,mayrose2005gamma,jia2014impact,kosakovsky2004simple,le2012modeling,jayaswal2014mixture}.
Although this significantly reduces the number of free parameters, choosing the number of the categories and specifying the corresponding statistical models remains rather arbitrary.

Furthermore, rate variation is likely not the most crucial aspect at which positions in a protein differ.
For the biological function of a protein, the properties of amino acids at different positions are crucial and experimentally measured amino-acid preferences at for each alignment column have been shown to greatly improve the phylogenetic fit \cite{bloom2014experimentally,bloom2014experimentally1}.



\section{Methods}
Most models of sequence evolution express the probability that sequence $\vec{s}$ evolved from sequence $\vec{r}$ in time $t$ as
\bq
P(\vec{s}\leftarrow \vec{r},t,\mat{Q}) = \prod_{a=1}^L \left(e^{\mat{Q}^{a} t}\right)_{s_{a},r_{a}}
\label{eq:P_prob}
\eq
where $\mat{Q}^{a}$ is the substitution matrix governing evolution at site $i$, and $s_{a}$ and $r_{a}$ are the sequence states at position $i$.
The product runs over all $L$ sites $a$ and amounts to assuming that different sites of the sequence evolve independently.

In absence of recombination, homologous sequences are related by a tree and the likelihood of observing an alignment $\mat{A}=\{\vec{s^k}, k=1..n\}$ conditional on the tree $T$ and the substitution model $\mat{Q}^{a}$ can be written in terms of propagators defined in Eq.~\ref{eq:P_prob}.
It is helpful to express this likelihood as product of sequence propagators defined in Eq.~\ref{eq:P_prob} between sequences at the ends of each branch in the tree (implicitly assuming that on different branches is independent and follows the same time reversible model).
Unknown sequences of internal nodes $\{\vec{s}'\}$ need to be summed over and the likelihood can be expressed as
\begin{equation}
	\LH(\mat{A}|T,\mat{Q}) = \sum_{\{\vec{s}'\}}\prod_{a=1}^L \eqp^{a}_{x^{0}_a} \prod_{k\in T}P(\vec{s}_{k_c}\leftarrow \vec{s}_{k_p},t,\mat{Q}) = \sum_{\{\vec{s}\}} e^{\lh(\{\vec{s}\} |T, \mat{Q})}  \ ,
	\label{eq:LH}
\end{equation}
where $\vec{s}_{k_c}$ and $\vec{s}_{k_p}$ are the child and parent sequences of branch $k$, respectively, and the factor $\prod \eqp^{a}_{x^{0}_a}$ is the product of the probabilities of the root sequence $s^{0}_a$ over all positions $i$.
The probabilities $\vec{\eqp}^{a}$ are the equiblirium probabilities of the substitution model at position $a$.
The latter ensures that the likelihood is insensitive to a particular choice of the tree root.
This equation defines the log-likelihood $\lh$ of a particular internal node assignment $\{\vec{s}'\}$ which is given by
\begin{equation}
	e^{\lh(\mat{A}, \{\vec{s}'\} |T, \mat{Q})} = \sum_a \left[\log(\eqp^a_{s^{0}_a}) + \sum_{k\in T} \log\left( e^{Q^at_k} \right)_{k^a_c, k^a_p}\right]
\end{equation}
where $k^a_c$ and $k^a_p$ are indices corresponding to the child and parent sequence of branch $k$.

The sum over unknown ancestral sequences can be computed efficiently using standard dynamic programming techniques.
Nevertheless, it requires $\mathcal{O}(n\times L \times q^2)$ operations (where $q$ is the size of the alphabet $\abet$) and optimizing it with respect to a large number of parameters is costly.
Our goal here is to infer site-specific substitution models using a computationally efficient iterative procedure.

Instead of inferring completely independent models for every site in the genome, we only allow for site specific rates and equilibrium frequencies while using the same transition matrix for every site.
Such a site-specific general time-reversible (GTR)) model, can be parameterized as:
\begin{eqnarray}
Q^{a}_{ij} &=& \mu^{a}\eqp^{a}_{i} W_{ij} \textrm{ for } i\neq j,\nonumber \\
Q^{a}_{ii} &=& -\sum_k Q^{a}_{ki}
\label{eq:Qij}
\end{eqnarray}
where $W_{ij}$ is a symmetric matrix with $W_{ii}=0$ and the second equation ensures conservation of probability.
In addition, we require $\sum_i \eqp^{a}_i = 1$ and $\sum_{a=1}^L\sum_{i\neq j}W_{ij}p^{a}_ip^{a}_j=L$ to ensure that the average rate per site is $\mu^{a}$.

% To regularize the parameter estimates when the data are not sufficiently informative to constrain them, we use Dirichlet and Gamma priors for $\eqp_i^a$ and $\mu^a$.
% \begin{equation}
% 	\log P(p_i^a) = C + (\alpha-1) \log \eqp_i^a \quad \mathrm{and} \quad \log P(\mu^a) = C -\frac{\beta \mu^a}{\bar{\mu}} + (\beta - 1)\log \mu^a
% \end{equation}

The derivatives of $\LH$ wrt to $\mu^a$, $\eqp_i^a$, and $W_{ij}$ need to vanish at the values that maximize $\LH$.
\begin{equation}
	\frac{\partial \LH(\mat{A}|T,\mat{Q})}{\partial X} = \sum_{\{\vec{s}\}} e^{\lh(\{\vec{s}\} |T, \mat{Q})} \frac{\partial \ell(\{\vec{s}\}|T,\mat{Q})}{\partial X} = 0
\end{equation}
where $X$ is one of the parameters we vary.

These conditions can be solved iteratively. We derive these update rule here $\mu^b$ and refer to the appendix for the other (more laborious) update rules.
\begin{equation}
	\frac{\partial \ell(\{\vec{s}\}|T,\mat{Q})}{\partial \mu^b} = \sum_{k\in T} \frac{t_k}{\mu^b}\frac{\sum_i Q^b_{k^b_c,i}\left( e^{Q^bt_k} \right)_{i, k^b_p}}{\left( e^{Q^bt_k} \right)_{k^b_c, k^b_p}}
\end{equation}
The individual terms in this sum behave very differently for cases where $k^b_c=k^b_p$ (no change at site $b$ on branch $k$) and $k^b_c\neq k^b_p$ (at least one mutation).
In the limit of short branches $\mu^b t_k\ll 1$, we can expand the matrix exponential $e^{\mat{Q}t}=\delta_{ij} + \mat{Q}t+\cdots$ to obtain more insightful expressions and we will separate branches with $k^b_c=k^b_p$  and $k^b_c\neq k^b_p$ right away
\begin{equation}
\begin{split}
	\frac{\partial \ell(\{\vec{s}\}|T,\mat{Q})}{\partial \mu^b} &\approx \sum_{k\in T, k^b_c=k^b_p} \frac{t_k}{\mu^b}\frac{\sum_i Q_{k^b,i}(\delta_{i,k^b} + t_k Q^b_{i,k^b}\cdots)}{1 - t_k\mu^b\sum_{j}\eqp_j W_{k^bj}+\ldots} +
	\sum_{k\in T, k^b_c\neq k^b_p} \frac{t_k}{\mu^b}\frac{\sum_i Q_{k^b_c,i}(\delta_{i,k^b_p} + t_k Q^b_{i,k^b_p}\cdots)}{t_k\mu^b\eqp_{k^b_c} W_{k^b_ck^b_p}+\ldots} \\
		&\approx -\sum_{k\in T, k^b_c=k^b_p}t_k \sum_i \eqp_{i}W_{i,k^b} +
	\sum_{k\in T, k^b_c\neq k^b_p} \frac{t_k}{\mu^b}\frac{p_{k^b_c}W_{k^b_c, l^b_p} + \cdots}{t_k\eqp_{k^b_c} W_{k^b_ck^b_p}+\ldots} \\
		&\approx -\sum_{k\in T, k^b_c=k^b_p}t_k \sum_i \eqp_{i}W_{i,k^b} + 	\sum_{k\in T, k^b_c\neq k^b_p} \frac{1}{\mu^b} \\
		&= -\sum_{ij}\eqp_{i}W_{i,j}\langle T_j^b\rangle + \frac{1}{\mu^b}\sum_{ij} n^b_{ij}
	\end{split}
\end{equation}
where $\langle T^b_j\rangle$ is the sum of all branch length along which site $b$ is in state $j$ and $n^b_{ij}$ is the number of times the sequences at site $b$ changes from $j$ to $i$ along branches of the tree.
This condition (and the corresponding ones in the supplement suggest solution at fixed $\langle T^b_j\rangle$ and $n^b_{ij}$ using the following iterative update:
\begin{equation}
\label{eq:update}
	\begin{split}
		\mu^b & \leftarrow \mu^b\frac{C+\sum_{ij} n^b_{ij}}{\sum_{ij}\mu^b \eqp_{i}W_{ij}\langle T_j^b\rangle} \\
		\eqp^b_i & \leftarrow \eqp^b_i\frac{C+\delta_{is^{b}} + \sum_{j\neq i} n^b_{ij}}{\eqp^b_i(qC+1) + \mu^b\eqp^b_i \sum_{j}W_{ij}\langle T_j^b\rangle} \\
		W_{ij} & \leftarrow W_{ij}\frac{\sum_b (n^b_{ij}+n^b_{ji})}{\sum_b \mu^b W_{ij}(\eqp_i^b\langle T_j^b\rangle + \eqp_j^b \langle T_i^b\rangle)} \\
	\end{split}
\end{equation}
where $C$ is a pseudocount analogous to a Dirichlet prior that  in absense of data will drive the $\eqp_i^b$ to a flat distribution and the substitution rates to $C$ over the total tree length.
To make the crucial features of these update rules more explicit, we have multiplied numerator and denominator with the parameter that is being updated. This (i) makes the update explicitly multiplicative and therefore ensures positivity, and (ii) illustrates that each of these rules are the ratio of the {\it observed} number of transitions between states $n^b_{ij}$ and the {\it expected} number $\eqp^b_i W_{ij}\langle T_j^b\rangle$ -- each appropriately summed over sites or states.

The validity of this iterative solution will depend on the accuracy of the linear approximation made, the degree to which the ancestral states can be reconstructed or summed over, and the accuracy of the tree reconstruction.
To assess these two sources of error independently, we simulated sequences evolving along a fixed tree and explicity recorded $n_{ij}^b$ and the $\langle T_j^b\rangle$ for a range of sample sizes, tree shapes, and branch lengths.
Fig.~\ref{fig:dressed} shows the accuracy of the inference as quantified by $\chi^2 = L^{-1}\sum_{a,i}(\hat{\eqp}_i^a - \eqp_i^a)^2$ where $\hat{\eqp}_i^a$ and $\hat{\eqp}_i^a$ the inferred and true equilibrium probabilities.


\begin{figure}[tb]
	\centering
	\includegraphics[width=0.48\textwidth]{../figures/p_dist_vs_treelength.pdf}
	\includegraphics[width=0.48\textwidth]{../figures/avg_rate_dressed.pdf}
	\caption{Accuracy of iterative estimation as a function of the total tree length, i.e., the expected number of state changes along the tree. (A) Mean squared error of the inferred $\eqp_i^a$ scales inversely with the tree length, suggesting the accuracy is limited by the number of observable mutations.
	The estimates are vastly more accurate than naive estimates from the state frequencies in alignment columns. Different curves are for $n\in [100,300,1000,3000]$ sequences.
	(B) While the $\eqp_i^a$ are accurately inferred by the linearized additive procedure with known internal states, the mutation rates are systematically underestimated, as expected. }
	\label{fig:dressed}
\end{figure}

Outside simulated sequences, the tree and the ancestral states at internal nodes are unknown and need to be reconstructed and summed over.
In the spirit of iterative updates, we first reconstruct those states using a simple evolutionary model such as Jukes-Cantor model, use the resulting ancestral reconstruction to estimate a site-specific model using Eq.~\ref{eq:update}, which can then be used to estimate an improved reconstruction.

Fig.~\ref{fig:reconstructed}A shows model inference accuracy for reconstructed trees (using iq-tree) and ancestral sequence inferred by TreeTime using a Jukes-Cantor model.
The estimates of $\eqp_i^a$ are about as accurate as those inferred from the true ancestral states und the root-to-tip distance reaches about 0.3 but then start to deteriorate and are eventually even worse than the naive estimates from the alignment.
This accuracy can be improved moderately by summing over uncertainty of the ancestral reconstruction.

A more substantial improvment in accuracy can be obtained when the inferred model is used to re-infer the ancestral states, followed by repeated model inference.
With this iterative procedure, model accuracy is comparable to that with known ancestral states up to root-to-tip distance of about 1.0.

In absense of strong regularization, maximum likelihood estimates of site specific equilibrium probabilities tend to `overfit' the data when the total number of observed mutations is low.
This is reflected in Fig.~\ref{fig:reconstructed}B showing the different in entropy $-\sum_i \eqp_i^a \log \eqp_i^a$ of the inferred and true distributions.
At low root-to-tip distances, the inferred distributions have too little entropy, meaning they are too concentrates on the few states that are observed.
This effect, however, is much less pronounced in the tree based inference compared to naive estimates directly from the alignment.
At very large root-to-tip distances, both the naive and the estimate based on true ancestral states become unbiased, while the model inference based on reconstructions results in broader $\eqp_i^a$ with higher entropy.


\begin{figure}[tb]
	\centering
	\includegraphics[width=0.48\textwidth]{../figures/p_dist_vs_rtt.pdf}
	\includegraphics[width=0.48\textwidth]{../figures/p_entropy_vs_rtt.pdf}
	\caption{Ancestral reconstruction: (A) At large root-to-tip distances, ancestral reconstruction becomes less and less accurate and estimation of equilibrium state frequencies fails. (B) For closely related sequences, the estimated equilibrium frequencies tend to be too concentrated in a few sites (low entropy), while at very large root-to-tip distances the estimates are too flat (high entropy). }
	\label{fig:reconstructed}
\end{figure}


\subsection*{Non-linear update rules}


\subsection*{Applications to large HIV alignments}



\bibliography{GTRbib}

\end{document}
