\documentclass[aps,rmp,twocolumn]{revtex4}
%\documentclass[a4paper,10pt]{scrartcl}
%\documentclass[aps,rmp,twocolumn]{revtex4}

\usepackage[utf8]{inputenc}
\usepackage{amsmath,graphicx}
\usepackage{color}
%\usepackage{cite}

\newcommand{\bq}{\begin{equation}}
\newcommand{\eq}{\end{equation}}
\newcommand{\bn}{\begin{eqnarray}}
\newcommand{\en}{\end{eqnarray}}
\newcommand{\Vadim}[1]{{\color{blue}Vadim: #1}}
\newcommand{\Richard}[1]{{\color{red}Richard: #1}}
\newcommand{\gene}[1]{{\it #1}}
\newcommand{\mat}[1]{{\bf #1}}
\newcommand{\vecb}[1]{{\bf #1}}
\newcommand{\abet}{\mathcal{A}}
\newcommand{\eqp}{p}
\newcommand{\LH}{\mathcal{L}}
\newcommand{\lh}{\ell}
\pdfinfo{%
  /Title    ()
  /Author   ()
  /Creator  ()
  /Producer ()
  /Subject  ()
  /Keywords ()
}


\begin{document}

\title{Site-specific substitution model inference with iterative tree likelihood maximization.}
\author{Vadim Puller$^{1,2,3}$ Pavel Sagulenko$^{1}$, Richard Neher$^{1,2,3}$}
\affiliation{$^{1}$Max Planck Institute for Developmental Biology, 72076 T\"ubingen, Germany\\
$^{2}$Biozentrum, University of Basel, Klingelbergstrasse 50/70, 4056 Basel, Switzerland\\
$^{3}$SIB Swiss Institute of Bioinformatics, Basel, Switzerland}

\date{\today}

\maketitle

\section*{Introduction}
Over time, genome sequences change through mutations and are reshuffeled by recombination.
Modifications to the genomes are filtered by selection for survival such that beneficial variants spread preferentially and those that impair function are purged.
As a result, some parts of genomes change rapidly, while other are strongly conserved.
In addition to variation of the evolutionary rate, different sites in a genome only explore different subsets of the available states.
Some positions in a protein, for example, might only allow for hydrophobic amino acids, while others require acidic side chains.
Patterns of conservation and variation, possibly involving more than one site, are therefore shaped by functional contraints which in turn allows inference of biological function from genetic variation.

Similarly, phylogenetics aims at reconstructing the relationships and history of homologous sequences from the substitutions that occurred in the past.
Modern phylogenetic methods describe this stochastic evolutionary process with probabilistic models of sequence evolution and aim to find phylogenies that either maximize the likelihood of observing the alignment or sample phylogenies from a posterior probability distribution \citep{felsenstein2004inferring}.

Inferring phylogenies is a computationally challenging problem since the number of phylogenies grows super-exponentially with the number of taxa and because the calculation of the likelihood is compulationally costly (though linear in the number of taxa).
Due to this computational complexity, the most commonly used substitution models are simple caricatures of biological complexity.
The simplest substitution models assume that all sites and sequence states are equivalent and evolve at the same rate, i.e., they assume an unconstrained non-functional sequence that mutates at random between the different sequence states.
Such simple models are clearly inadequate and ignoring rate heterogeneity tends to result, among other issues, in biased estimates of divergence times \citep{yang1996among}.
Most commonly used models account for variation in substitution rates among sites and average properties of the substitution process such as transition/transversion bias or more likely substitution between similar amino acids \citep{yang_maximum_1994,FastTree2,nguyen_iq-tree:_2015,stamatakis_raxml_2014}.
To avoid over-fitting, these methods typically don't estimate rates for each site, but treat site-specific rates as random effects that are integrated out (often using discrete approximations of a Gamma distribution \citep{yang1996among}, mixture of multiple uni-model distributions \citep{mayrose2005gamma}, or a small number of fixed rates).

In addition to rate variation, different sites in a protein differ in amino acid they allow.
Recent deep mutational scanning experimentes have shown that site specific preferences are mostly conserved between moderately diverged proteins \citep{doud_site-specific_2015}.
Using such experimentally inferred site specific models in phylogenetic inference greatly improves the likelihood of the data \cite{bloom2014experimentally}.
More than two decades ago, \citet{halpern_evolutionary_1998} pointed out that ignoring that equilibrium frequencies vary from one position to another will result in underestimation of branch lengths of a phylogeny -- possibly dramatically when frequencies are heavily skewed.
\citet{hilton_modeling_2018} recently showed that experimentally measured preference not only improve the phylogenetic fit, but also results in longer branch length estimates.

On the otherhand, estimating such models, also known as mutation-selection balance models \citep{bruno1996modeling,yang2008mutation}, from data exacerbates the over-fitting problem.
In the context of the site-specific models this issue has become known as {\em extensive parametrization} or even {\em infinitely many parameters problem} \cite{Rodrigue557,spielman2016extensively}.
With sufficient data, however, site specific parameters can be accurately estimated \citep{tamuri_estimating_2012,spielman2016extensively,scheffler2014validity}.
Here, we implement an EM-style algorithm inspired by \citep{bruno1996modeling} to infer site-specific rates and preferences from simulated data, quantify its accuracy and the different sources of bias and noise, and show how divergence time estimates depend on model deviations.

\begin{figure*}[htb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/p_dist_vs_treelength_nuc_ratealpha3.0}.pdf}
	%\includegraphics[width=0.32\textwidth]{../figures/avg_rate_dressed.pdf}
	\includegraphics[width=0.48\textwidth]{{../figures/rate_correlation_dressed_nuc_ratealpha3.0}.pdf}
	\caption{{\bf Accuracy of iterative estimation of site-specific GTR models as a function of the expected number of state changes along the tree.} (A) Mean squared error of the inferred $\eqp_i^a$ scales inversely with the tree length, suggesting the accuracy is limited by the Poisson statistics of observable mutations.
	The estimates are vastly more accurate than naive estimates from the state frequencies in alignment columns. Different curves are for $n\in [100,300,1000]$ sequences.
	% (B) While the $\eqp_i^a$ are accurately inferred by the linearized additive procedure with known internal states, the mutation rates are systematically underestimated, as expected.
	(B) The relative subsitution rates are accurately inferred as soon as the typical site experiences several substitutions across the tree as quantified here as Pearson correlation coefficient between true and inferred rates. Regularization via pseudo-counts reduces over-fitting at low divergence.  }
	\label{fig:dressed}
\end{figure*}


%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Model
%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Efficient inference of site specific substitution models}
Following previous authors \citet{halpern_evolutionary_1998}, we assume that the substitution matrix $W_{ij}$ between states $j,i$ does not depend on the position in the sequence, while
the overall rates and the equilibrium frequencies depend on the site.
Such a site-specific general time-reversible (GTR)) model, can be parameterized as:
\begin{eqnarray}
Q^{a}_{ij} &=& \mu^{a}\eqp^{a}_{i} W_{ij} \textrm{ for } i\neq j,\nonumber \\
Q^{a}_{ii} &=& -\sum_k Q^{a}_{ki}
\label{eq:Qij}
\end{eqnarray}
where $W_{ij}$ is a symmetric matrix with $W_{ii}=0$ and the second equation ensures conservation of probability.
In addition, we require $\sum_i \eqp^{a}_i = 1$ and $\sum_{a=1}^L\sum_{i\neq j}W_{ij}p^{a}_ip^{a}_j=L$ to ensure that the average rate per site is $\mu^{a}$.

Extending the iterative expectation maximization approach by \citet{bruno1996modeling}, we show in the appendix that the iterative update rules
\begin{equation}
\label{eq:update}
	\begin{split}
		\mu^b & \leftarrow \mu^b\frac{C+\sum_{ij} n^b_{ij}}{\sum_{ij}\mu^b \eqp_{i}W_{ij}\tau_j^b} \\
		\eqp^b_i & \leftarrow \eqp^b_i\frac{C+\delta_{is^{b}} + \sum_{j\neq i} n^b_{ij}}{\eqp^b_i(qC+1) + \mu^b\eqp^b_i \sum_{j}W_{ij}\tau_j^b} \\
		W_{ij} & \leftarrow W_{ij}\frac{\sum_b (n^b_{ij}+n^b_{ji})}{\sum_b \mu^b W_{ij}(\eqp_i^b\tau_j^b + \eqp_j^b \tau_i^b)} \\
	\end{split}
\end{equation}
approximately maximize the likelihood of the data.
Here $\tau^b_j$ is the time site $b$ spends in state $j$ across the tree and $n^b_{ij}$ are the number of transitions from state $j$ to state $i$ at site $b$
$C$ is a pseudocount analogous to a Dirichlet prior that  in absense of data will drive the $\eqp_i^b$ to a flat distribution and the substitution rates to $C$ over the total tree length.
To make the behavior of these update rules more explicit, we have multiplied numerator and denominator with the parameter that is being updated. This (i) makes the update explicitly multiplicative and therefore ensures positivity, and (ii) illustrates that each of these rules are the ratio of the {\it observed} number of transitions between states $n^b_{ij}$ and the {\it expected} number $\eqp^b_i W_{ij}\tau_j^b$ -- each appropriately summed over sites or states.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% ACCURACY
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Accuracy of inferences}
The validity of this iterative solution will depend on the accuracy of the linear approximation made, the degree to which $n_{ij}^b$ and $\tau_{j}^b$ can be estimated, and the accuracy of the tree reconstruction.
In addition, the most likely model parameters will deviate from the true parameters due to over-fitting.

To assess these sources of error independently, we simulated sequences evolving along a fixed tree and explicity recorded $n_{ij}^b$ and the $\tau_j^b$ for a range of sample sizes, levels of divergence, and different degrees of variation between sites.
We quantify the accuracy of the inference as $\chi^2 = L^{-1}\sum_{a,i}(\hat{\eqp}_i^a - \eqp_i^a)^2$ where $\hat{\eqp}_i^a$ and $\eqp_i^a$ are the inferred and true equilibrium probabilities and compare inference using (i) the true ancestral sequences, (ii) reconstructed ancestral sequences, and (iii) reconstructions on the true phylogenetic tree.



\begin{figure*}[tb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/p_dist_vs_rtt_nuc_ratealpha3.0}.pdf}
	\includegraphics[width=0.48\textwidth]{{../figures/p_dist_vs_rtt_aa_ratealpha3.0}.pdf}
	\caption{{\bf Quantification of errors stemming from tree inference and ancestral reconstruction.}
	The left and right panels show the mean-squared deviation $\chi^2$ of inferred $\hat{\eqp}_i^a$ from the true $\eqp_i^a$ for 4-letter and 20-letter alphabets, respectively.
	At large root-to-tip distances, ancestral reconstruction becomes less and less certain and estimation of $\eqp_i^a$ fails (red lines).
	These errors are gradually eliminated by first summing over ancestral uncertainty (violet), iteratively redoing ancestral reconstruction using the inferred model (brown), and re-optimizing branch length using the updated models (or using the true tree, yellow/pink).
	Data in this figure uses was generated assuming Gamma distributed rate variation with $\alpha=3$.}
	\label{fig:reconstructed}
\end{figure*}


Fig.~\ref{fig:dressed} shows the average squared error $\chi^2$ for a trees with $n\in [100,300,1000,3000]$ taxa and a range of substitution rates as a function of the expected number of substitutions per site for estimates of $\eqp_i^a$ from the alignment and using the iterative scheme with correct $n_{ij}^a$ and $\tau_i^a$ as input.
The plot is shown on double logarithmic scales, such that the approximately straight decrease of $\chi^2$ with slope 1 implies that the accuracy is inversely proportional to the expected number of substitutions.
This scaling suggests that accuracy is limited by the inherent stochasticity of the evolutionary process and that the inference uses all available information efficiently.
At the largest substitution rates, branch length are on the order of 0.5 and the linear approximation underlying the iterative equations is not longer accurate.
Nevertheless, the acccuracy of the $\hat{\eqp}_i^a$ continues to increase.
By contrast, using the frequencies of different sequences states in the alignment as estimates for $\eqp_i^b$ is much less accurate due to the correlation induced by shared ancestry.

The accuracy of substitution rates, however, is affected by linearization and we consistently underestimate $\mu^b$.
This is expected as the linearization ignores cases where the same site changes twice along a branch in very much the same way as Hamming distance will underestimate branch length.
The relative subsitution rates, however, are accurately estimated (with Pearson correlation coefficients $>0.9$ as soon as the majority of sites experience several mutations across the tree), see Fig.~\ref{fig:dressed}B.

%%% effect of reconstruction
In a typical scenario, ancestral sequences are unknown and need to be inferred or summed over (marginalized)
To test the influence of tree and ancestral state reconstruction, we built phylogenetic trees using IQ-tree \citep{nguyen_iq-tree:_2015} with a GTR+R10 model or FastTree \citep{Price} using the default 20 category model for simulated nucleotide and amino-acid sequences, respectively.
Fig~\ref{fig:reconstructed} compares different schemes to reconstruct ancestral sequences, infer substitution models, and optimize the tree.
The simplest approach is to take the inferred tree as given, reconstruct the ancestral states a standard evolutionary model (e.g.~Jukes-Cantor model) and calculate $n^b_{ij}$ and $\tau_i^b$ from this reconstruction.
This naive approach works well up to root-to-tip distances of about 0.3, beyond which estimates deteriorate.
Instead of using only the most likely ancestral sequences, we can instead average over all possible ancestral states, which results in a modest improvemet in accurary.
More significant gains are made when iterating model inference and ancestral reconstruction using the inferred site specific model.
The accuracy now continues to improve up to root-to-tip distances of about one.
At this level of divergence, tree reconstruction starts becoming problematic and branch length deviate substantially from their true values.
Using the true tree instead of the reconstructed tree leads to continuous improvements of accuracy with increasing levels of divergence.
Similar improvements are achieved by optimizing tree branch lengths along with the model.


\subsection*{Ignoring site specific frequencies results in dramatic branch length underestimates}
As previously observed by various authors \citep{halpern_evolutionary_1998,hilton_modeling_2018}, sites with heavily skewed preferences for specific states result in underestimation of branch lengths if uniform equilibrium frequencies are assumed.
This is a straightforward consequence of the fact that the probability of observing the same state at random is $\sum_i{\eqp_i^a}^2$ is increasing sharply with more peaked preferences.
Models that don't account for site specific frequencies will take a large number of sites that agree between two sequences as evidence for their close evolutionary relationship, while it is simply a consequence of resampling the same states with high probability.
In other words, this underestimation is the result of incorrect models of saturation.

\begin{figure*}[tb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/2019-06-01_simulatedData_L1000_ratealpha1.5_alpha1.0_nuc_length}.pdf}
	\includegraphics[width=0.48\textwidth]{{../figures/2019-06-01_simulatedData_L1000_ratealpha1.5_alpha1.0_nuc_depth}.pdf}
	\caption{{\bf Skewed equilibrium concentration results in branch length under estimates}.
	Panel A shows the inferred average branch length as estimated by IQ-Tree and TreeTime as a function of the true average branch length. Panel B shows the results of the same optimization for the average root-to-tip distance, which is dominated by deep long branches which are more strongly affected by underestimation. Not that inferred site specific models only partially ameliorate underestimation, see main text. }
	\label{fig:tree_length}
\end{figure*}


%% dependence on pseudo count
Fig.~\ref{fig:tree_length} shows the average branch length (panel A) and the average root-to-tip distance of mid-point rooted tree (panel B) with branch length optimized using (i) the true model, (ii) using the GTR+R10 model of IQ-tree, and (iii) inferred models using different degrees of regularization.
While branch length are accurately estimated when using the true model, they are systematically underestimated with increasing pseudo-counts and or without site specificity.
Underestimation is particularly severe for the root-to-tip distance which is dominated by long branches deep in the phylogeny which are particularly prone to underestimation.
When ignoring site-specific preferences, the inferred root-to-tip distance is essentially independent from the true value for distance greater than 1 (Fig.~\ref{fig:tree_length}B).
This effect is entirely due to skewed equilbrium frequencies, as branch length inference by IQ-tree is accurate if the simulated data had flat $\eqp_i^a=q^{-1}$.
%% dependence on nucleotide and amino acid skew
The underestimation of branch lengths is less of a problem when using larger alphabets and the true probability are not too heavily skewed.

\begin{figure}[tb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/model_deviation_2019-06-01_simulatedData_L1000_ratealpha1.5_alpha1.0_nuc_n100_m0.35}.pdf}
	\caption{{\bf Moderate model inaccuracies result in large substantial branch length underestimation}.
	}
	\label{fig:model_deviation}
\end{figure}

Surprisingly, the inferred site specific model only partially rectify the problem of branch length under-estimation, despite the fact that these models are close to the true model in terms of low $\chi^2$.
The principle contributor to this deviation are inaccuracies in the rate estimates and combining the true rates with inferred preferences largely eliminates the error in branch length estimation.
If, however, preferences are estimated poorly for example through excessive regularization ($\pc=10$ in Fig.~\ref{fig:tree_length}), using the correct rates hardly improves the estimate.
The latter is consistent with the observation that standard ML tree inference using IQ-tree with a GTR+R10 model severely underestimates branch lengths.

\subsection*{Minor model deviations can have substantial effects on divergence estimates}
In the previous section, we have observed that even small model deviations, for example those introduced by strong regularization, already result in substantial mis-estimation.
To more systematically investigate this effect, we constructed mixtures of the true model and a model with flat $\eqp_i^a$ and/or $\mu^a$.

Fig.~\ref{fig:model_deviation}A shows how deviation from the true model affect relative branch length estimates.
Deviation in rate estimates result immediately in underestimated branch length, while substantial effect of too flat site specific preferences only manifest themselves once deviations are of order 0.3.
If the same preference are used for every site, however, deviations are substantial.
Deviations in rate and preferences are approximately additive.
These observations are consistent with the finding above using inferred preferences and true substitution rates result in much more accurate branch length estimates than inferring both rates and preferences.

Fig.~\ref{fig:model_deviation}B shows the degree of mis-estimation when using flat preferences, flat rates, or both as a function of the degree of divergence.
The relative error increases approximately linearly with the average evolutionary rate.


\subsection*{Applications to large HIV alignments}
Since its zoonosis, HIV-1 has diversified into several different subtypes that differ from each other at about 20\% of sites in their genome.
For each subtype, thousands of sequences are available, which typically differ by about 10\% \citep{LANL}.
This large sample of moderately diverged sequences is a unique opportunity to apply our iterative inference frame work.
We downloaded alignments of HIV-1 pol sequences from the LANL data base, constructed phylogenetic trees, and inferred site specific GTR models at the nucleotide and amino acid level.

One of the original motivations for site-specific models was the idea that equilibrium frequencies reflect the relative fitness of the different states at each position \citep{halpern_evolutionary_1998}.
Simple population genetic models predict that the fixation probability $f_{ij}^a$ of a mutation from state $j$ to state $i$ at site $a$ on the fitness different $s^a_{ij}$ and the effective population size $N$ as
\begin{equation}
	f^a_{ij} = \frac{1-e^{-2s^a_{ij}}}{1-e^{-2Ns^a_{ij}}} \approx
	\begin{cases}
		2s^a_{ij} & s^a_{ij}>0 \\
		2s^a_{ij} e^{2Ns^a_{ij}} & s^a_{ij} < 0
	\end{cases}
\end{equation}
On longer time scales, the fixation rates $f^a_{ij}$ should correspond to transitions rates of the site specific GTR model $\mu^a Q^a_{ij}$.
The effective population size $N$ plays the role of a coalescent time scale $T_c$ and in general has little to do with census population sizes, in particular in rapidly adapting populations.
Nevertheless, the logarithm of the ratio $\log f^a_{ij}/f^a_{ji} \approx 2Ns^a_{ij}$ is an interpretable quantity related to the average fitness difference between states on time scales longer than the population genetic scale $N\sim T_c$.
We generalize this notion to multiple states and define a fitness score of state $i$ at position $a$ as the ratio of rates into and out of state $i$
\begin{equation}
	e^{2N s_i^a} = \frac{\Gamma_{in}}{\Gamma_{out}} = \frac{\sum_j \pi_i^a W_{ij}}{\sum_j W_{ij}\pi_j^a}
\end{equation}
The logarithm of this ratio is expected to be proportional to the fitness difference between state $i$ and the average alternative state at site $a$, while the multiplicative factor $2N$ is unknown.
We compare this proxy of fitness differences to fitness costs measured using mutation selection balance models and with-in host diversity data of HIV \citep{zanini2017vivo}.
However, instead of a linear relation ship between $s_i^a$ and within host estimates of fitness costs, we observe a linear relationship between the logarithm of fitness effects measured with-in host and $s_i^a$ (
see Fig.~\ref{fig:HIV_fitness})
This relationship explains about half the variance of nucleotide effects and slightly less of amino acid effects.

The fact that this relationship deviates from the `cross-sectional$\sim$ with-in host' expectation and instead is approximately `cross-sectional $\sim \log$ with-in host' points towards different process that drive cross-sectional and with-in host diversification.
Within host fitness effects are masked and damped at the population level, possibly due to fluctuating by diverse host immune systems or epistasis \citep{zanini2015population,shekhar_spin_2013}.
The ratio of in/out rates $e^{-2Ns_i^a}$ is, however, a very good correlate of alignment diversity (see Fig.~\ref{fig:HIV_fitness}).

\begin{figure*}[tb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/HIV_B_pol_fitness_pc_0.001}.pdf}
	\includegraphics[width=0.48\textwidth]{{../figures/HIV_B_pol_fitness_pc_0.001_aa}.pdf}
	\caption{{\bf Intra-host vs cross-sectional mutation selection balance.}
	Panel A\&B shows the ratio of in/out rates for consensus nucleotides/amino acids along the \gene{pol} of HIV-1 subtype B vs of fitness costs of non-consensus states estimated from within-host mutation selection balance.
	The logarithm of the rate ratio is roughly linear in the logarithm of the fitness cost.}
	\label{fig:HIV_fitness}
\end{figure*}




\section*{Discussion}
Sequence evolution models attempt to capture some biological realism while being computationally tractable and easy to parameterize.
A common compromise is to model rate heterogeneity as random effects, while assuming identical site preferences across the sequence.
Different positions in a DNA sequence or protein, however, differ markedly in the preferences for different states -- this site specific preference is the basis of all common homology search tools.
While the need to account for this variation has been known for decades \citep{bruno1996modeling,halpern1998evolutionary}, it is typically not modeled due to over-fitting concerns and due to computational complexity.
However, amino acid preferences can now be measured with high-throughput \citep{fowler_deep_2014} and data sets are increasing rapidly such that more complex models can be estimated.
Large data sets, however, require efficient and scalable methods.

Here, we introduce and implement a method to estimate site specific GTR models using an iterative EM-stype approach inspired by \citet{bruno1996modeling}.
We show that site specific models can be inferred with an accuracy limited by the Poisson statistics of the substitution process using an efficient iterative scheme.
For short branch length a parsimonious ancestral reconstruction is sufficient, while for more diverged samples and iterative inference of the model, the ancestral states, and the branch lengths is necessary.
The computational complexity of the inference is dominated by ancestral reconstruction and branch length optimization, which is quadratic in the alphabet size and linear in sample size and sequence length \citep{felsenstein2004inferring}.
We have implemented the algorithm for site-specific model inference and branch lengths optimization in TreeTime \citep{sagulenko2017treetime}.


We further explored how model choice affects the maximum likelihood estimates of branch length.
As has been reported before, branch length are underestimated when variation in rate or preferences are not fully accounted for \citep{halpern1998evolutionary,hilton_modeling_2018}.
While such underestimation often has a moderate effect on the total length of a tree, it can result in substantial underestimation of root-to-tip distance that are dominated by a few long branches close to the root.
While using the correct model results in correct branch length estimates, joint inference of site-specific models and branch lengths is typically unable to recover the true branch lengths and substantial underestimation remains.
This mirrors observations by \citet{hilton_modeling_2018}, who found that preferences measured for influenza HA proteins of type H1 or H3 affect branch length of in the vicinity of the focal sequence, but not globally on the tree.

The potentially large error in branch length estimating can seriously affect deep divergence time estimates.
Typically, short branches close to the tips of the tree are used to calibrate molecular clock models.
The deep branches, however, tend to be much longer and rate estimation and variation of site specific preferences will result in saturation effects not accounted for by the model \citep{hilton_modeling_2018}.
This effective variation in rate is related to the effect of transient deleterious mutations that inflate the rate on short time scales \citep{ho_time_2005}: the time it takes purifying selection to prune deleterious mutations is related to the relaxation time scales of GTR models with site specific preferences.
Instead of the $q-1$ degenerate eigenvalues of a Jukes-Cantor model, each site has a spectrum of eigenvalues and different eigenmodes relax at different speeds, generating apparently time dependent rates.

While accurate estimates of the site-specific GTR models are a possible given a large enough set of mildly diverged sequences, joint infererence of model and branch lengths remains difficult.
This difficulty can be largely traced down to poorly constrained estimates of site-specific rates -- one primary reason why these rates are typically not estimated but treated as random effects \citep{yang1996among}.





\begin{itemize}
	\item underestimation of long branches, dubious deep time trees
	\item relation to frequent reversions
	\item fitness inference, functional constraints
	\item caveat about interactions
\end{itemize}

\bibliography{GTRbib}

\onecolumngrid

\section*{Materials and Methods}
\subsection*{Model and notation}
\label{sec:MM_model}
Most models of sequence evolution express the probability that sequence $\vec{s}$ evolved from sequence $\vec{r}$ in time $t$ as
\bq
P(\vec{s}\leftarrow \vec{r},t,\mat{Q}) = \prod_{a=1}^L \left(e^{\mat{Q}^{a} t}\right)_{s_{a},r_{a}}
\label{eq:P_prob}
\eq
where $\mat{Q}^{a}$ is the substitution matrix governing evolution at site $i$, and $s_{a}$ and $r_{a}$ are the sequence states at position $i$.
The product runs over all $L$ sites $a$ and amounts to assuming that different sites of the sequence evolve independently.

In absence of recombination, homologous sequences are related by a tree and the likelihood of observing an alignment $\mat{A}=\{\vec{s^k}, k=1..n\}$ conditional on the tree $T$ and the substitution model $\mat{Q}^{a}$ can be written in terms of propagators defined in Eq.~\ref{eq:P_prob}.
It is helpful to express this likelihood as product of sequence propagators defined in Eq.~\ref{eq:P_prob} between sequences at the ends of each branch in the tree (implicitly assuming that on different branches is independent and follows the same time reversible model).
Unknown sequences of internal nodes $\{\vec{s}'\}$ need to be summed over and the likelihood can be expressed as
\begin{equation}
	\LH(\mat{A}|T,\mat{Q}) = \sum_{\{\vec{s}'\}}\prod_{a=1}^L \eqp^{a}_{x^{0}_a} \prod_{k\in T}P(\vec{s}_{k_c}\leftarrow \vec{s}_{k_p},t,\mat{Q}) = \sum_{\{\vec{s}\}} e^{\lh(\{\vec{s}\} |T, \mat{Q})}  \ ,
	\label{eq:LH}
\end{equation}
where $\vec{s}_{k_c}$ and $\vec{s}_{k_p}$ are the child and parent sequences of branch $k$, respectively, and the factor $\prod \eqp^{a}_{x^{0}_a}$ is the product of the probabilities of the root sequence $s^{0}_a$ over all positions $i$.
The probabilities $\vec{\eqp}^{a}$ are the equiblirium probabilities of the substitution model at position $a$.
The latter ensures that the likelihood is insensitive to a particular choice of the tree root.
This equation defines the log-likelihood $\lh$ of a particular internal node assignment $\{\vec{s}'\}$ which is given by
\begin{equation}
	e^{\lh(\mat{A}, \{\vec{s}'\} |T, \mat{Q})} = \sum_a \left[\log(\eqp^a_{s^{0}_a}) + \sum_{k\in T} \log\left( e^{Q^at_k} \right)_{k^a_c, k^a_p}\right]
\end{equation}
where $k^a_c$ and $k^a_p$ are indices corresponding to the child and parent sequence of branch $k$.

The sum over unknown ancestral sequences can be computed efficiently using standard dynamic programming techniques.
Nevertheless, it requires $\mathcal{O}(n\times L \times q^2)$ operations (where $q$ is the size of the alphabet $\abet$) and optimizing it with respect to a large number of parameters is costly.
Our goal here is to infer site-specific substitution models using a computationally efficient iterative procedure.

Instead of inferring completely independent models for every site in the genome, we follow \citet{halpern_evolutionary_1998} and only allow for site specific rates and equilibrium frequencies while using the same transition matrix for every site.
Such a site-specific general time-reversible (GTR)) model, can be parameterized as:
\begin{eqnarray}
Q^{a}_{ij} &=& \mu^{a}\eqp^{a}_{i} W_{ij} \textrm{ for } i\neq j,\nonumber \\
Q^{a}_{ii} &=& -\sum_k Q^{a}_{ki}
\label{eq:Qij}
\end{eqnarray}
where $W_{ij}$ is a symmetric matrix with $W_{ii}=0$ and the second equation ensures conservation of probability.
In addition, we require $\sum_i \eqp^{a}_i = 1$ and $\sum_{a=1}^L\sum_{i\neq j}W_{ij}p^{a}_ip^{a}_j=L$ to ensure that the average rate per site is $\mu^{a}$.

% To regularize the parameter estimates when the data are not sufficiently informative to constrain them, we use Dirichlet and Gamma priors for $\eqp_i^a$ and $\mu^a$.
% \begin{equation}
% 	\log P(p_i^a) = C + (\alpha-1) \log \eqp_i^a \quad \mathrm{and} \quad \log P(\mu^a) = C -\frac{\beta \mu^a}{\bar{\mu}} + (\beta - 1)\log \mu^a
% \end{equation}

\subsection*{Iterative optimization algorithm}
The derivatives of $\LH$ wrt to $\mu^a$, $\eqp_i^a$, and $W_{ij}$ need to vanish at the values that maximize $\LH$.
\begin{equation}
	\frac{\partial \LH(\mat{A}|T,\mat{Q})}{\partial X} = \sum_{\{\vec{s}\}} e^{\lh(\{\vec{s}\} |T, \mat{Q})} \frac{\partial \ell(\{\vec{s}\}|T,\mat{Q})}{\partial X} = 0
\end{equation}
where $X$ is one of the parameters we vary.

These conditions can be solved iteratively. We derive these update rule here $\mu^b$ and refer to the appendix for the other (more laborious) update rules.
\begin{equation}
	\frac{\partial \ell(\{\vec{s}\}|T,\mat{Q})}{\partial \mu^b} = \sum_{k\in T} \frac{t_k}{\mu^b}\frac{\sum_i Q^b_{k^b_c,i}\left( e^{Q^bt_k} \right)_{i, k^b_p}}{\left( e^{Q^bt_k} \right)_{k^b_c, k^b_p}}
\end{equation}
The individual terms in this sum behave very differently for cases where $k^b_c=k^b_p$ (no change at site $b$ on branch $k$) and $k^b_c\neq k^b_p$ (at least one mutation).
In the limit of short branches $\mu^b t_k\ll 1$, we can expand the matrix exponential $e^{\mat{Q}t}=\delta_{ij} + \mat{Q}t+\cdots$ to obtain more insightful expressions and we will separate branches with $k^b_c=k^b_p$  and $k^b_c\neq k^b_p$ right away
\begin{equation}
\begin{split}
	\frac{\partial \ell(\{\vec{s}\}|T,\mat{Q})}{\partial \mu^b} &\approx \sum_{k\in T, k^b_c=k^b_p} \frac{t_k}{\mu^b}\frac{\sum_i Q_{k^b,i}(\delta_{i,k^b} + t_k Q^b_{i,k^b}\cdots)}{1 - t_k\mu^b\sum_{j}\eqp_j W_{k^bj}+\ldots} +
	\sum_{k\in T, k^b_c\neq k^b_p} \frac{t_k}{\mu^b}\frac{\sum_i Q_{k^b_c,i}(\delta_{i,k^b_p} + t_k Q^b_{i,k^b_p}\cdots)}{t_k\mu^b\eqp_{k^b_c} W_{k^b_ck^b_p}+\ldots} \\
		&\approx -\sum_{k\in T, k^b_c=k^b_p}t_k \sum_i \eqp_{i}W_{i,k^b} +
	\sum_{k\in T, k^b_c\neq k^b_p} \frac{t_k}{\mu^b}\frac{p_{k^b_c}W_{k^b_c, l^b_p} + \cdots}{t_k\eqp_{k^b_c} W_{k^b_ck^b_p}+\ldots} \\
		&\approx -\sum_{k\in T, k^b_c=k^b_p}t_k \sum_i \eqp_{i}W_{i,k^b} + 	\sum_{k\in T, k^b_c\neq k^b_p} \frac{1}{\mu^b} \\
		&= -\sum_{ij}\eqp_{i}W_{i,j}\tau_j^b + \frac{1}{\mu^b}\sum_{ij} n^b_{ij}
	\end{split}
\end{equation}
where $\tau^b_j$ is the sum of all branch length along which site $b$ is in state $j$ and $n^b_{ij}$ is the number of times the sequences at site $b$ changes from $j$ to $i$ along branches of the tree.
This condition (and the corresponding ones in the supplement suggest solution at fixed $\tau^b_j$ and $n^b_{ij}$ using the following iterative update:
\begin{equation}
\label{eq:update}
	\begin{split}
		\mu^b & \leftarrow \mu^b\frac{C+\sum_{ij} n^b_{ij}}{\sum_{ij}\mu^b \eqp_{i}W_{ij}\tau_j^b} \\
		\eqp^b_i & \leftarrow \eqp^b_i\frac{C+\delta_{is^{b}} + \sum_{j\neq i} n^b_{ij}}{\eqp^b_i(qC+1) + \mu^b\eqp^b_i \sum_{j}W_{ij}\tau_j^b} \\
		W_{ij} & \leftarrow W_{ij}\frac{\sum_b (n^b_{ij}+n^b_{ji})}{\sum_b \mu^b W_{ij}(\eqp_i^b\tau_j^b + \eqp_j^b \tau_i^b)} \\
	\end{split}
\end{equation}
where $C$ is a pseudocount analogous to a Dirichlet prior that in absense of data will drive the $\eqp_i^b$ to a flat distribution and the substitution rates to $C$ over the total tree length.
To make the crucial features of these update rules more explicit, we have multiplied numerator and denominator with the parameter that is being updated. This (i) makes the update explicitly multiplicative and therefore ensures positivity, and (ii) illustrates that each of these rules are the ratio of the {\it observed} number of transitions between states $n^b_{ij}$ and the {\it expected} number $\eqp^b_i W_{ij}\tau_j^b$ -- each appropriately summed over sites or states.


\subsection*{HIV sequence analysis}
HIV-1 sequences were downloaded from LANL HIV database \citep{LANL} setting filters to ``one sequence per patient'', ``non-ACGT\textless 0.3''.
Separate downloads where made for sequences the following sets: \gene{pol}, subtype B (2019-05-05); \gene{pol}, subtype C (2019-05-13); \gene{gag}, \gene{nef}, subtype B (2019-06-07).

Sequences were aligned to the HXB2 reference sequences using mafft \citep{katoh2013mafft}, phylogenies were inferred using IQ-tree \citep{nguyen_iq-tree:_2015}, and ancestral sequences were inferred using TreeTime \citep{sagulenko2017treetime} via the nextstrain's augur pipeline \citep{hadfield_nextstrain:_2018}.
The different steps were assembled into a pipeline using the workflow manager Snakemake \citep{koster_snakemakescalable_2012}.

The sequences and trees were then used for site specific GTR inference as implemented in TreeTime \citep{sagulenko2017treetime}.
The scripts detailing this analysis and producing the figures are available on GitHub in repository \href{https://github.com/neherlab/SiteSpecificGTR}{neherlab/SiteSpecificGTR}.

\clearpage
\appendix
\renewcommand{\thefigure}{S\arabic{figure}}
\renewcommand{\thetable}{S\arabic{table}}
\setcounter{figure}{0}
\setcounter{table}{0}

\section*{Supplementary figures}
\begin{figure*}[htb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/p_dist_vs_treelength_aa_ratealpha3.0}.pdf}
	%\includegraphics[width=0.32\textwidth]{../figures/avg_rate_dressed.pdf}
	\includegraphics[width=0.48\textwidth]{{../figures/rate_correlation_dressed_aa_ratealpha3.0}.pdf}
	\caption{Accuracy of iterative estimation as a function of the total tree length, i.e., the expected number of state changes along the tree, for amino acid alphabets. (A) Mean squared error of the inferred $\eqp_i^a$ scales inversely with the tree length, suggesting the accuracy is limited by the number of observable mutations.
	The estimates are vastly more accurate than naive estimates from the state frequencies in alignment columns. Different curves are for $n\in [100,300,1000]$ sequences.
	% (B) While the $\eqp_i^a$ are accurately inferred by the linearized additive procedure with known internal states, the mutation rates are systematically underestimated, as expected.
	(C) The relative subsitution rates are accurately inferred as soon as the typical site experiences several substitutions across the tree as quantified here as Pearson correlation coefficient between true and inferred rates. Regularization via pseudo-counts reduces over-fitting at low divergence.  }
	\label{fig:dressed}
\end{figure*}

\begin{figure*}[tb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/HIV_B_gag_fitness_pc_0.001}.pdf}
	\includegraphics[width=0.48\textwidth]{{../figures/HIV_B_nef_fitness_pc_0.001}.pdf}
	\caption{{\bf Intra-host vs cross-sectional mutation selection balance.}
	In analogy to Fig.~\ref{fig:HIV_fitness}, the panels A\&B show the ratio of in/out rates for consensus nucleotides acids along the \gene{gag} and \gene{nef} gene of HIV-1 subtype B vs of fitness costs of non-consensus states estimated from within-host mutation selection balance.
	The logarithm of the rate ratio is roughly linear in the logarithm of the fitness cost.}
	\label{fig:HIV_fitness}
\end{figure*}


\begin{figure*}[htb]
	\centering
	\includegraphics[width=0.48\textwidth]{{../figures/HIV_B_pol_alignment_div_pc_0.001}.pdf}
	\includegraphics[width=0.48\textwidth]{{../figures/HIV_B_pol_alignment_div_pc_0.001_aa}.pdf}
	\caption{{\bf Diversity and GTR rate estimates.}
	Panel A\&B shows the ratio of in/out rates for consensus nucleotides/amino acids along the \gene{pol} of HIV-1 subtype B vs the diversity in columns of the corresponding alignment.}
	\label{fig:HIV_diversity}
\end{figure*}





\end{document}



